### ISSUE TITLE
Typechecking result depends of using temporary variable with 'print' function

### ISSUE THREAD

**Issue by Renegatto** [python/mypy#8792]
### The bugreport

A code that required to reproduce a bug:
```python
from typing import TypeVar, Callable, Union, Iterable, List
T = TypeVar('T')

def find(predicate: Callable[[T],bool], xs:Iterable[T]) -> Union[T,None]:
    for x in xs:
        if predicate(x):
            return x
    return None

foo : List[int] = [] #any numbers, any count of elements

def bar(x: int) -> bool: return True

```
### A bug:
The results of several cases to **print** results of `find(bar, foo)` (lines 14, 15):
1 ❌ (unexpected mypy output here)
```python
print(find(bar, foo))
```
Mypy's output:
> test.py:14: **error**: Argument 1 to **"find1"** has incompatible type **"Callable[[int], bool]"**; expected **"Callable[[object], bool]**"

2 ✅
```python
baz = find(bar, foo)
print(baz)
```
Mypy's output:
> **Success: no issues found in 1 source file**

3 ✅
```python
baz : Union[int, None] = find(bar, foo)
print(baz)
```
Mypy's output:
> **Success: no issues found in 1 source file**

⚠️ a type `int` in this code can be replaced to another type (i tried `str`) but the bug still will be reproduce
⚠️ replacing `Union[T,None]` to `Optional[T]` doesnt change anything

* What is the actual behavior/output?
Different typecheck results of cases that equal enough.
* What is the behavior/output you expect?
No issues found, or at least same results of typechecking of all cases
* What are the versions of mypy and Python you are using?
mypy ver. 0.770
python ver. 3.7.3
* Do you see the same issue after installing mypy from Git master?
I didnt/wont try to do it.
* What are the mypy flags you are using? (For example --strict-optional)
None.

**Comment by AlexWaygood** [comment]
The mypy error is [not reproducible](https://mypy-play.net/?mypy=0.930&python=3.10&flags=warn-unreachable&gist=52f2f7aec02175cabcf9222749ea7eab) on 0.930+.
